protocol/x/clob/keeper/liquidations.go

```diff
@@ -664,63 +700,82 @@ func (k Keeper) GetLiquidationInsuranceFundDelta(
 
 // GetPerpetualPositionToLiquidate determines which position and position size to liquidate on the
 // passed-in subaccount (after accounting for the `update`). It will return the perpetual id that
-// will be used for liquidating the perpetual position and the number of quantums to liquidate
-// from the perpetual position (positive if long, negative if short).
+// will be used for liquidating the perpetual position
 // This function returns an error if the subaccount has no perpetual positions to liquidate.
 func (k Keeper) GetPerpetualPositionToLiquidate(
 	ctx sdk.Context,
 	subaccountId satypes.SubaccountId,
 ) (
 	perpetualId uint32,
-	quantums *big.Int,
 	err error,
 ) {
 	// Fetch the subaccount from state.
 	subaccount := k.subaccountsKeeper.GetSubaccount(ctx, subaccountId)
 	subaccountLiquidationInfo := k.GetSubaccountLiquidationInfo(ctx, subaccountId)
 
-	var perpetualPosition *satypes.PerpetualPosition
+	numPositions := len(subaccount.PerpetualPositions)
+	indexOffset := k.GetPseudoRand(ctx).Intn(numPositions)
 
-	for _, position := range subaccount.PerpetualPositions {
+	for i := 0; i < numPositions; i++ {
+		position := subaccount.PerpetualPositions[(i+indexOffset)%numPositions]
 		// Note that this could run in O(n^2) time. This is fine for now because we have less than a hundred
 		// perpetuals and only liquidate once per subaccount per block. This means that the position with smallest
 		// id will be liquidated first.
 		if !subaccountLiquidationInfo.HasPerpetualBeenLiquidatedForSubaccount(position.PerpetualId) {
-			perpetualPosition = position
-			break
+			return position.PerpetualId, nil
 		}
 	}
 
 	// Return an error if there are no perpetual positions to liquidate.
-	if perpetualPosition == nil {
-		return 0,
-			nil,
+	return 0,
+		errorsmod.Wrapf(
+			types.ErrNoPerpetualPositionsToLiquidate,
+			"Subaccount ID: %v",
+			subaccount.Id,
+		)
+}
+
+// GetLiquidatablePositionSizeDelta returns the max number of base quantums to liquidate
+// from the perpetual position without exceeding the block and position limits.
+func (k Keeper) GetLiquidatablePositionSizeDelta(
+	ctx sdk.Context,
+	subaccountId satypes.SubaccountId,
+	perpetualId uint32,
+) (
+	deltaQuantums *big.Int,
+	err error,
+) {
+	subaccount := k.subaccountsKeeper.GetSubaccount(ctx, subaccountId)
+	perpetualPosition, exists := subaccount.GetPerpetualPositionForId(perpetualId)
+	if !exists {
```

Comment:
IMO this should be a panic, since I think [this logic](https://github.com/dydxprotocol/v4-chain/blob/1a8bafb3b294df1188db5c879435d0486167fa62/protocol/x/clob/abci.go#L188) in `PrepareCheckState` will panic on this case. Up to you

Reply:
then we can just panic upstream, right?